### DISCUSSION TITLE
webextension or wasm to crypt/decrypt with pkcs#11 hsm

### DISCUSSION THREAD

**Discussion by Cyrille37** [OpenSC/OpenSC#2839] [General]
Hello

I'm discovering the HSM world and be able to use HSM (SafeNet 5110) in Firefox for authentification, thanks a lot to the crypto community :-)

But I'm browsing the web to find how to use the HSM to crypt and decrypt data in the browser (Firefox or Chromium) but I couldn't find anything :-( Are you aware about some works around pcks#11 webextension or wasm for crypt and decrypt in a web browser ?

[This post](https://stackoverflow.com/a/48357920/4833788) tell it's not possible without an external process ...

Thanks & cheers

**Comment by Cyrille37** [comment]
Thales share a Python API to talk with theirs libraries for SafeNet eTokens.

[ThalesGroup/pycryptoki](https://github.com/ThalesGroup/pycryptoki/blob/master/docs/getting_started.rst)

May we do the same, but with WASM talk with OpenSC or SAC libraries ?

[Export native API to WASM application](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/export_native_api.md)

I'll keep digging...

**Comment by Cyrille37** [comment]
Another way could be to use the webextension [Native Messaging](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging) to talk with OpenSC.

I'll keep digging...

**Comment by frankmorgner** [comment]
See https://github.com/OpenSC/OpenSC/discussions/2908#discussioncomment-7618664
